casbin/distributed_enforcer.py

```diff
@@ -0,0 +1,134 @@
+        """
+        if(should_persist):
+            try:
+                if(isinstance(self.adapter, batch_adapter)):
+                    self.adapter.remove_policy(sec, ptype, rules)
+            except Exception as e:
+                self.logger.log("An exception occurred: " + e)
+
+        self.get_model().remove_policies(sec, ptype, rules)
+
+        if sec == "g":
+            try:
+                self.build_incremental_role_links(PolicyOp.Policy_remove, ptype, rules)
+            except Exception as e:
+                self.logger.log("An exception occurred: " + e)
+                return rules
+
+        return rules
+
+    def remove_filtered_policy_self(self, should_persist, sec, ptype, field_index, *field_values):
+        """
+        remove_filtered_policy_self provides a method for dispatcher to remove an authorization
+        rule from the current policy,field filters can be specified.
+        The function returns the rules affected and error.
+        """
+        if should_persist:
+            try:
+                self.adapter.remove_filtered_policy(sec, ptype, field_index, field_values)
+            except Exception as e:
+                self.logger.log("An exception occurred: " + e)
+
+        effects = self.get_model().remove_filtered_policy_returns_effects(sec, ptype, field_index, field_values)
+
+        if sec == "g":
+            try:
+                self.build_incremental_role_links(PolicyOp.Policy_remove, ptype, effects)
+            except Exception as e:
+                self.logger.log("An exception occurred: " + e)
+                return effects
+
+        return effects
+
+    def clear_policy_self(self, should_persist):
+        """
+        clear_policy_self provides a method for dispatcher to clear all rules from the current policy.
+        """
+        if should_persist:
+            try:
+                self.adapter.save_policy(None)
+            except Exception as e:
+                self.logger.log("An exception occurred: " + e)
+
+        self.get_model().clear_policy()
+
+    def update_policy_self(self, should_persist, sec, ptype, old_rule, new_rule):
+        """
+        update_policy_self provides a method for dispatcher to update an authorization rule from the current policy.
+        """
+        if should_persist:
+            try:
+                if isinstance(self.adapter, update_adapter):
+                    self.adapter.update_policy(sec, ptype, old_rule, new_rule)
+            except Exception as e:
+                self.logger.log("An exception occurred: " + e)
+                return False
+
+        rule_updated = self.get_model().update_policy(sec, ptype, old_rule, new_rule)
+
+        if not rule_updated:
+            return False
+
+        rules = []
+        if sec == "g":
+            try:
+                rules.append(old_rule)
+                self.build_incremental_role_links(PolicyOp.Policy_remove, ptype, rules)
```

Comment:
Actully `build_increment_role_links` takes a 2-d list as an argument. `old_rule` is a 1-d list. So what I am doing here is I am taking an empty list `rules`, appending old_rule in rules and passing rules to `build_increment_role_links`. 

To remove it I will have to add line : `self.build_incremental_role_links(PolicyOp.Policy_remove, ptype, [old_rule])`. 

Please tell me if I should change that line to the above line or previous one is fine.

Reply:
You are right. I made a mistake, because I looked at [`UpdatePoliciesSelf`](https://github.com/casbin/casbin/blob/86d48b589606a67f8dc3c51486e8daa1b743f579/enforcer_distributed.go#L148), while reviewing your code.

Nonetheless you should change that line to `self.build_incremental_role_links(PolicyOp.Policy_remove, ptype, [old_rule])`.
The current implementation recreates the old role link at line 129.

![Screenshot_2021-03-25_11-53-42](https://user-images.githubusercontent.com/8122139/112462511-6b82a400-8d61-11eb-9fcb-5c780aafd72c.png)

Reply:
Ok I'll make changes.